File: src/main/java/li/klass/fhem/backup/ImportExportService.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.backup;

import android.content.Context;

import com.google.common.base.Charsets;
import com.google.common.base.MoreObjects;
import com.google.common.base.Optional;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import com.google.gson.Gson;

import net.lingala.zip4j.core.ZipFile;
import net.lingala.zip4j.exception.ZipException;
import net.lingala.zip4j.exception.ZipExceptionConstants;
import net.lingala.zip4j.model.ZipParameters;
import net.lingala.zip4j.util.Zip4jConstants;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.Map;

import javax.inject.Inject;

import li.klass.fhem.connection.backend.ConnectionService;
import li.klass.fhem.devices.favorites.backend.FavoritesService;
import li.klass.fhem.service.NotificationService;
import li.klass.fhem.util.ApplicationProperties;
import li.klass.fhem.util.CloseableUtil;
import li.klass.fhem.util.ReflectionUtil;
import li.klass.fhem.util.io.FileSystemService;
import li.klass.fhem.util.preferences.SharedPreferencesService;

import static com.google.common.base.Preconditions.checkArgument;

public class ImportExportService {
    public enum ImportStatus {
        SUCCESS, INVALID_FILE, WRONG_PASSWORD
    }

    public static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormat.forPattern("yyyy-MM-dd_HH-mm");
    public static final String SHARED_PREFERENCES_FILE_NAME = "sharedPreferences.json";
    private Logger LOGGER = LoggerFactory.getLogger(ImportExportService.class);

    @Inject
    SharedPreferencesService sharedPreferencesService;

    @Inject
    FileSystemService fileSystemService;

    @Inject
    FavoritesService favoritesService;

    @Inject
    ApplicationProperties applicationProperties;

    @Inject
    public ImportExportService() {
    }

    protected Map<String, String> getSharedPreferencesExportKeys(Context context) {
        ImmutableMap.Builder<String, String> builder = ImmutableMap.<String, String>builder()
                .put("CONNECTIONS", ConnectionService.Companion.getPREFERENCES_NAME())
                .put("NOTIFICATIONS", NotificationService.PREFERENCES_NAME)
                .put("DEFAULT", applicationProperties.getApplicationSharedPreferencesName(context));
        for (String preferenceName : favoritesService.getPreferenceNames(context)) {
            builder.put("FAVORITE_" + preferenceName, preferenceName);
        }
        return builder.build();
    }

    public File exportSettings(String password, Context context) {
        Map<String, Map<String, ?>> toExport = Maps.newHashMap();

        for (Map.Entry<String, String> exportValue : getSharedPreferencesExportKeys(context).entrySet()) {
            Map<String, ?> values = sharedPreferencesService.listAllFrom(exportValue.getValue(), context);
            toExport.put(exportValue.getKey(), toExportValues(values));
        }

        return createZipFrom(toExport, Optional.fromNullable(password));
    }

    public Map<String, String> toExportValues(Map<String, ?> values) {
        Map<String, String> toExport = Maps.newHashMap();
        for (Map.Entry<String, ?> entry : values.entrySet()) {
            toExport.put(entry.getKey(), entry.getValue() + "/" + entry.getValue().getClass().getName());
        }
        return toExport;
    }


    public Map<String, ?> toImportValues(Map<String, String> values) {
        Map<String, Object> toImport = Maps.newHashMap();
        for (Map.Entry<String, String> entry : values.entrySet()) {
            int separator = entry.getValue().lastIndexOf("/");
            Class<?> clazz = ReflectionUtil.classForName(entry.getValue().substring(separator + 1));
            String value = entry.getValue().substring(0, separator);
            toImport.put(entry.getKey(), typedValueFor(value, clazz));
        }
        return toImport;
    }

    private Object typedValueFor(String value, Class<?> type) {
        if (type.isAssignableFrom(Integer.class)) {
            return Integer.parseInt(value);
        } else if (type.isAssignableFrom(Float.class)) {
            return Float.parseFloat(value);
        } else if (type.isAssignableFrom(Boolean.class)) {
            return Boolean.parseBoolean(value);
        } else if (type.isAssignableFrom(String.class)) {
            return value;
        } else if (type.isAssignableFrom(Double.class)) {
            return Double.parseDouble(value);
        } else {
            throw new IllegalArgumentException("don't know how to handle " + type.getName());
        }
    }

    public boolean isValidZipFile(File file) {
        try {
            ZipFile zipFile = new ZipFile(file);
            return zipFile.isValidZipFile();
        } catch (ZipException e) {
            LOGGER.error("error while reading zip file", e);
            return false;
        }
    }

    public boolean isEncryptedFile(File file) {
        try {
            ZipFile zipFile = new ZipFile(file);
            checkArgument(zipFile.isValidZipFile());
            return zipFile.isEncrypted();
        } catch (ZipException e) {
            LOGGER.error("error while reading zip file", e);
            throw new IllegalArgumentException(e);
        }
    }

    @SuppressWarnings("unchecked")
    public ImportStatus importSettings(File file, String password, Context context) {
        ZipFile zipFile;
        InputStreamReader importReader = null;

        try {
            zipFile = new ZipFile(file);
            if (zipFile.isEncrypted()) {
                zipFile.setPassword(MoreObjects.firstNonNull(password, ""));
            }

            if (zipFile.getFileHeader(SHARED_PREFERENCES_FILE_NAME) == null) {
                return ImportStatus.INVALID_FILE;
            }

            zipFile.extractFile(SHARED_PREFERENCES_FILE_NAME, fileSystemService.getCacheDir(context).getAbsolutePath());

            importReader = new InputStreamReader(new FileInputStream(new File(fileSystemService.getCacheDir(context), SHARED_PREFERENCES_FILE_NAME)));
            Map<String, Map<String, String>> content = new Gson().fromJson(importReader, Map.class);
            for (Map.Entry<String, Map<String, String>> entry : content.entrySet()) {
                sharedPreferencesService.writeAllIn(getSharedPreferencesExportKeys(context).get(entry.getKey()), toImportValues(entry.getValue()), context);
            }
            return ImportStatus.SUCCESS;

        } catch (ZipException e) {
            LOGGER.error("importSettings(" + file.getAbsolutePath() + ") - cannot import", e);
            if (e.getCode() == ZipExceptionConstants.WRONG_PASSWORD || e.getMessage().contains("Wrong Password")) {
                return ImportStatus.WRONG_PASSWORD;
            } else {
                return ImportStatus.INVALID_FILE;
            }
        } catch (Exception e) {
            LOGGER.error("importSettings(" + file.getAbsolutePath() + ") - cannot import", e);
            return ImportStatus.INVALID_FILE;
        } finally {
            CloseableUtil.close(importReader);
        }
    }

    private File createZipFrom(Map<String, Map<String, ?>> toExport, Optional<String> password) {

        ByteArrayInputStream stream = null;
        try {
            String exportedJson = new Gson().toJson(toExport);

            File exportFile = new File(getExportDirectory(), getBackupFileName());
            LOGGER.info("export file location is {}", exportFile.getAbsolutePath());
            ZipFile zipFile = new ZipFile(exportFile);


            ZipParameters parameters = new ZipParameters();
            parameters.setCompressionMethod(Zip4jConstants.COMP_DEFLATE);
            parameters.setFileNameInZip(SHARED_PREFERENCES_FILE_NAME);
            parameters.setSourceExternalStream(true);
            if (password.isPresent()) {
                parameters.setEncryptFiles(true);
                parameters.setEncryptionMethod(Zip4jConstants.ENC_METHOD_STANDARD);
                parameters.setPassword(password.get());
            }

            stream = new ByteArrayInputStream(exportedJson.getBytes(Charsets.UTF_8));
            zipFile.addStream(stream, parameters);

            return exportFile;
        } catch (ZipException e) {
            LOGGER.error("cannot create zip", e);
            throw new IllegalStateException(e);
        } finally {
            CloseableUtil.close(stream);
        }
    }

    private String getBackupFileName() {
        return "andFHEM-" + DATE_TIME_FORMATTER.print(DateTime.now()) + ".backup";
    }

    public File getExportDirectory() {
        return fileSystemService.getOrCreateDirectoryIn(fileSystemService.getDocumentsFolder(), "andFHEM");
    }
}

File: app/src/main/java/li/klass/fhem/backup/ImportExportService.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.backup

import android.content.Context
import com.google.gson.Gson
import li.klass.fhem.connection.backend.ConnectionService
import li.klass.fhem.devices.list.favorites.backend.FavoritesService
import li.klass.fhem.service.NotificationService
import li.klass.fhem.util.ApplicationProperties
import li.klass.fhem.util.CloseableUtil
import li.klass.fhem.util.Logging
import li.klass.fhem.util.ReflectionUtil
import li.klass.fhem.util.io.FileSystemService
import li.klass.fhem.util.preferences.SharedPreferencesService
import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.exception.ZipException
import net.lingala.zip4j.model.ZipParameters
import net.lingala.zip4j.model.enums.CompressionMethod
import net.lingala.zip4j.model.enums.EncryptionMethod
import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat
import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.InputStreamReader
import javax.inject.Inject

class ImportExportService @Inject constructor(
        private val sharedPreferencesService: SharedPreferencesService,
        private val fileSystemService: FileSystemService,
        private val favoritesService: FavoritesService,
        private val applicationProperties: ApplicationProperties
) : Logging {
    private val backupFileName: String
        get() = "andFHEM-" + dateTimeFormatter.print(DateTime.now()) + ".backup"

    val exportDirectory: File
        get() = fileSystemService.getOrCreateDirectoryIn(fileSystemService.documentsFolder, "andFHEM")

    enum class ImportStatus {
        SUCCESS, INVALID_FILE, WRONG_PASSWORD
    }

    private fun getSharedPreferencesExportKeys(context: Context): Map<String, String> {
        val builder = mutableMapOf(
                "CONNECTIONS" to ConnectionService.PREFERENCES_NAME,
                "NOTIFICATIONS" to NotificationService.PREFERENCES_NAME,
                "DEFAULT" to applicationProperties.getApplicationSharedPreferencesName(context)
        )
        for (preferenceName in favoritesService.getPreferenceNames()) {
            builder["FAVORITE_$preferenceName"] = preferenceName
        }
        return builder.toMap()
    }

    fun exportSettings(password: String?, context: Context): File {
        val toExport = mutableMapOf<String, Map<String, *>>()

        for ((key, value) in getSharedPreferencesExportKeys(context)) {
            val values = sharedPreferencesService.listAllFrom(value)
            toExport[key] = toExportValues(values)
        }

        return createZipFrom(toExport.toMap(), password)
    }

    fun toExportValues(values: Map<String, *>) = values.entries
            .map { it.key to it.value.toString() + "/" + (it.value as Any).javaClass.name }
            .toMap()


    fun toImportValues(values: Map<String, String>): Map<String, *> {
        val toImport = mutableMapOf<String, Any>()
        for ((key, value1) in values) {
            val separator = value1.lastIndexOf("/")
            val clazz = ReflectionUtil.classForName(value1.substring(separator + 1))
            val value = value1.substring(0, separator)
            toImport[key] = typedValueFor(value, clazz)
        }
        return toImport
    }

    private fun typedValueFor(value: String, type: Class<*>): Any {
        return if (type.isAssignableFrom(Int::class.java) || type.isAssignableFrom(Integer::class.java)) {
            Integer.parseInt(value)
        } else if (type.isAssignableFrom(Float::class.java) || type.isAssignableFrom(java.lang.Float::class.java)) {
            java.lang.Float.parseFloat(value)
        } else if (type.isAssignableFrom(Boolean::class.java) || type.isAssignableFrom(java.lang.Boolean::class.java)) {
            java.lang.Boolean.parseBoolean(value)
        } else if (type.isAssignableFrom(String::class.java)) {
            value
        } else if (type.isAssignableFrom(Double::class.java) || type.isAssignableFrom(java.lang.Double::class.java)) {
            java.lang.Double.parseDouble(value)
        } else if (type.isAssignableFrom(Long::class.java) || type.isAssignableFrom(
                        java.lang.Long::class.java)) {
            java.lang.Long.parseLong(value)
        } else {
            throw IllegalArgumentException("don't know how to handle " + type.name)
        }
    }

    fun toZipFile(file: File): ZipFile? {
        return try {
            ZipFile(file).takeIf {
                it.isValidZipFile
            }
        } catch (e: ZipException) {
            logger.error("error while reading zip file ${file.absolutePath}", e)
            null
        }
    }

    fun importSettings(zipFile: ZipFile, password: String?, context: Context): ImportStatus {
        try {
            if (zipFile.isEncrypted) {
                zipFile.setPassword((password ?: "").toCharArray())
            }

            if (zipFile.getFileHeader(SHARED_PREFERENCES_FILE_NAME) == null) {
                return ImportStatus.INVALID_FILE
            }

            zipFile.extractFile(SHARED_PREFERENCES_FILE_NAME, fileSystemService.getCacheDir(context).absolutePath)

            val content = InputStreamReader(FileInputStream(File(fileSystemService.getCacheDir(context), SHARED_PREFERENCES_FILE_NAME)))
                    .use { reader ->
                        Gson().fromJson<Map<String, Map<String, String>>>(reader, Map::class.java)
                    }

            importNonFavorites(context, content)
            importFavorites(context, content)

            return ImportStatus.SUCCESS

        } catch (e: ZipException) {
            logger.error("importSettings(${zipFile.file.absolutePath}) - cannot import", e)
            return if (e.type == ZipException.Type.WRONG_PASSWORD || (e.message
                            ?: "").contains("Wrong Password")) {
                ImportStatus.WRONG_PASSWORD
            } else {
                ImportStatus.INVALID_FILE
            }
        } catch (e: Exception) {
            logger.error("importSettings(${zipFile.file.absolutePath}) - cannot import", e)
            return ImportStatus.INVALID_FILE
        }
    }

    private fun importNonFavorites(context: Context, content: Map<String, Map<String, String>>) =
            import(context, content) { !it.startsWith("FAVORITE")}

    private fun importFavorites(context: Context, content: Map<String, Map<String, String>>) =
            import(context, content) { it.startsWith("FAVORITE")}


    private fun import(context: Context, content: Map<String, Map<String, String>>, predicate: (String) -> Boolean) {
        val exportKeys = getSharedPreferencesExportKeys(context)
        content.entries
                .filter { predicate(it.key) }
                .filter { exportKeys.containsKey(it.key) }
                .map { exportKeys.getValue(it.key) to toImportValues(it.value) }
                .forEach { sharedPreferencesService.writeAllIn(it.first, it.second) }
    }

    private fun createZipFrom(toExport: Map<String, Map<String, *>>, password: String?): File {

        var stream: ByteArrayInputStream? = null
        try {
            val exportedJson = Gson().toJson(toExport)

            val exportFile = File(exportDirectory, backupFileName)
            logger.info("export file location is {}", exportFile.absolutePath)
            val zipFile = ZipFile(exportFile, password?.toCharArray())


            val parameters = ZipParameters()
            parameters.compressionMethod = CompressionMethod.DEFLATE
            parameters.fileNameInZip = SHARED_PREFERENCES_FILE_NAME
            if (password != null) {
                parameters.isEncryptFiles = true
                parameters.encryptionMethod = EncryptionMethod.ZIP_STANDARD
            }

            stream = ByteArrayInputStream(exportedJson.toByteArray(Charsets.UTF_8))
            zipFile.addStream(stream, parameters)

            return exportFile
        } catch (e: ZipException) {
            logger.error("cannot create zip", e)
            throw IllegalStateException(e)
        } finally {
            CloseableUtil.close(stream)
        }
    }

    companion object {
        private val dateTimeFormatter = DateTimeFormat.forPattern("yyyy-MM-dd_HH-mm")!!
        private const val SHARED_PREFERENCES_FILE_NAME = "sharedPreferences.json"
    }
}
